cmd/legacy/legacy.go

```diff
@@ -155,31 +158,55 @@ func (l LegacyFileSet) GetBlock(ts int64) (goDB.AggFlowMap, error) {
 	}
 
 	for i := 0; i < len(protoBlock); i++ {
+		sip := rawIPToAddr(sipBlock[i*16 : i*16+16])
+		dip := rawIPToAddr(dipBlock[i*16 : i*16+16])
+		if sip.Is4() != dip.Is4() && !sip.IsUnspecified() {
+			logrus.StandardLogger().Warnf("source / destination IP v4 / v6 mismatch: %s / %s, will convert to IPv6\n", sip, dip)
+		}
 
-		var K goDB.Key
-		var V goDB.Val
+		var K types.Key
+		var V types.Val
 
-		copy(K.Sip[:], sipBlock[i*16:i*16+16])
-		copy(K.Dip[:], dipBlock[i*16:i*16+16])
-		copy(K.Dport[:], dportBlock[i*2:i*2+2])
-		K.Protocol = protoBlock[i]
+		if sip.Is4() && dip.Is4() {
+			K = types.NewV4KeyStatic(sip.As4(), dip.As4(), dportBlock[i*2:i*2+2], protoBlock[i])
+		} else {
+			K = types.NewV6KeyStatic(sip.As16(), dip.As16(), dportBlock[i*2:i*2+2], protoBlock[i])
+		}
 
 		V.NBytesRcvd = binary.BigEndian.Uint64(bytesRcvdBlock[i*8 : i*8+8])
 		V.NBytesSent = binary.BigEndian.Uint64(bytesSentBlock[i*8 : i*8+8])
 		V.NPktsRcvd = binary.BigEndian.Uint64(pktsRcvdBlock[i*8 : i*8+8])
 		V.NPktsSent = binary.BigEndian.Uint64(pktsSentBlock[i*8 : i*8+8])
 
-		entry, exists := data[K]
-		if exists {
-			entry.NBytesRcvd += V.NBytesRcvd
-			entry.NBytesSent += V.NBytesSent
-			entry.NPktsRcvd += V.NPktsRcvd
-			entry.NPktsSent += V.NPktsSent
-			data[K] = entry
-		} else {
-			data[K] = &V
-		}
+		data.SetOrUpdate(K, V.NBytesRcvd, V.NBytesSent, V.NPktsRcvd, V.NPktsSent)
 	}
 
 	return data, nil
 }
+
+func rawIPToAddr(ip []byte) netip.Addr {
+	zeros := numZeros(ip)
+	ind := len(ip)
+	if zeros == 12 {
+		// only read first 4 bytes (IPv4)
+		ind = 4
+	}
+	netIP, ok := netip.AddrFromSlice(ip[:ind])
+	if !ok {
+		return netip.Addr{}
+	}
+	return netIP
+}
+
+func numZeros(ip []byte) uint8 {
```

Comment:
Jup, but only for the conversion tool (that's why I moved it here so it can be removed after the release of v4). The old files still use our padded encoding, so this is required to detect it from the source blocks.